mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -868,3 +907,93 @@ impl ChecksumConfig {
         &self.inner
     }
 }
+
+/// Checksum algorithm.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ChecksumAlgorithm {
+    /// No checksums.
+    None,
+    /// Crc32c checksum.
+    Crc32c,
+    /// Crc32 checksum.
+    Crc32,
+    /// Sha1 checksum.
+    Sha1,
+    /// Sha256 checksum.
+    Sha256,
+}
+
+impl From<aws_s3_checksum_algorithm> for ChecksumAlgorithm {
+    fn from(algorithm: aws_s3_checksum_algorithm) -> Self {
+        match algorithm {
+            aws_s3_checksum_algorithm::AWS_SCA_NONE => ChecksumAlgorithm::None,
+            aws_s3_checksum_algorithm::AWS_SCA_CRC32C => ChecksumAlgorithm::Crc32c,
+            aws_s3_checksum_algorithm::AWS_SCA_CRC32 => ChecksumAlgorithm::Crc32,
+            aws_s3_checksum_algorithm::AWS_SCA_SHA1 => ChecksumAlgorithm::Sha1,
+            aws_s3_checksum_algorithm::AWS_SCA_SHA256 => ChecksumAlgorithm::Sha256,
+            _ => unreachable!("unknown aws_s3_checksum_algorithm"),
+        }
+    }
+}
+
+impl From<ChecksumAlgorithm> for aws_s3_checksum_algorithm {
+    fn from(algorithm: ChecksumAlgorithm) -> Self {
+        match algorithm {
+            ChecksumAlgorithm::None => aws_s3_checksum_algorithm::AWS_SCA_NONE,
+            ChecksumAlgorithm::Crc32c => aws_s3_checksum_algorithm::AWS_SCA_CRC32C,
+            ChecksumAlgorithm::Crc32 => aws_s3_checksum_algorithm::AWS_SCA_CRC32,
+            ChecksumAlgorithm::Sha1 => aws_s3_checksum_algorithm::AWS_SCA_SHA1,
+            ChecksumAlgorithm::Sha256 => aws_s3_checksum_algorithm::AWS_SCA_SHA256,
+        }
+    }
+}
+
+/// Info for the caller to review before an upload completes.
+#[derive(Debug)]
+pub struct UploadReview {
+    /// Info about each part uploaded.
+    pub parts: Vec<UploadPartReview>,
+    /// The checksum algorithm used.
+    pub checksum_algorithm: ChecksumAlgorithm,
+}
+
+impl UploadReview {
+    fn new(review: &aws_s3_upload_review) -> Self {
+        let checksum_algorithm = review.checksum_algorithm.into();
+        let count = review.part_count;
+        let mut parts = Vec::new();
+        for i in 0..count {
+            // SAFETY: `part_array` is an array of length `count`.
+            let part = unsafe { &*review.part_array.add(i) };
+            parts.push(UploadPartReview::new(part));
+        }
+        Self {
+            parts,
+            checksum_algorithm,
+        }
+    }
+
+    /// Failure error code.
+    pub const FAILURE: i32 = aws_s3_errors::AWS_ERROR_S3_CANCELED as i32;
+}
+
+/// Info about a single part, for the caller to review before the upload completes.
+#[derive(Debug)]
+pub struct UploadPartReview {
+    /// Size in bytes of this part.
+    pub size: u64,
+
+    /// Checksum string (usually base64-encoded). This is empty if no checksum is used.
+    pub checksum: String,
```

Comment:
Again should probably be `Option<String>` rather than exposing the "empty byte cursor" logic